
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../01_6_nfcore_intro/">
      
      
        <link rel="next" href="../01_8_nfcore_config/">
      
      
        
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>1.7 Running nf-core on HPC - Nextflow on HPC</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-TNWWV3HK57"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-TNWWV3HK57",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-TNWWV3HK57",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#17-running-nf-core-on-hpc" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Nextflow on HPC" class="md-header__button md-logo" aria-label="Nextflow on HPC" data-md-component="logo">
      
  <img src="../../assets/usyd-logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Nextflow on HPC
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              1.7 Running nf-core on HPC
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Sydney-Informatics-Hub/nextflow-hpc-workshop" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    Sydney-Informatics-Hub/nextlow-hpc-workshop
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Nextflow on HPC" class="md-nav__button md-logo" aria-label="Nextflow on HPC" data-md-component="logo">
      
  <img src="../../assets/usyd-logo.png" alt="logo">

    </a>
    Nextflow on HPC
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Sydney-Informatics-Hub/nextflow-hpc-workshop" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    Sydney-Informatics-Hub/nextlow-hpc-workshop
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../setup/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Set up
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Part 1
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Part 1
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_0_intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1.0 Introduction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_1_hpc_for_workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1.1 HPC for bioinformatics workflows
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_2_hpc_software/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1.2 Software is different on HPC
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_3_hpc_architecture/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1.3 HPC architecture
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_4_smarter/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1.4 Work smarter, not harder!
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_5_nf_hpc/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1.5 Running Nextflow on HPC
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_6_nfcore_intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1.6 Intro to nf-core
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    1.7 Running nf-core on HPC
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    1.7 Running nf-core on HPC
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#171-an-overview-of-the-inputs" class="md-nav__link">
    <span class="md-ellipsis">
      
        1.7.1 An overview of the inputs
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.7.1 An overview of the inputs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastq-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        FASTQ files
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fasta-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        FASTA files
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bwa-index" class="md-nav__link">
    <span class="md-ellipsis">
      
        BWA Index
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#172-write-a-simple-run-script" class="md-nav__link">
    <span class="md-ellipsis">
      
        1.7.2 Write a simple run script
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#173-not-quite-ready" class="md-nav__link">
    <span class="md-ellipsis">
      
        1.7.3 Not quite ready!
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_8_nfcore_config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1.8 Configuring nf-core
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_9_nfcore_layer/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1.9 Layering Nextflow configurations
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_10_outro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    1.10 Summary
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Part 2
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Part 2
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../part2/20_intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2.0 Introduction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../part2/21_firstrun/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2.1 Running a custom pipeline on HPC
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../part2/22_reporting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2.2 Pipeline monitoring and reporting
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../part2/23_strategies/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2.3 Optimising Nextflow for HPC
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../part2/24_resourcing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2.4 Assigning process resources
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../part2/25_parallelisation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2.5 Parallelisation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../part2/27_scale/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2.6 Scale to multiple samples
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../part2/28_outro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2.7 Summary
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#171-an-overview-of-the-inputs" class="md-nav__link">
    <span class="md-ellipsis">
      
        1.7.1 An overview of the inputs
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.7.1 An overview of the inputs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastq-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        FASTQ files
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fasta-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        FASTA files
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bwa-index" class="md-nav__link">
    <span class="md-ellipsis">
      
        BWA Index
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#172-write-a-simple-run-script" class="md-nav__link">
    <span class="md-ellipsis">
      
        1.7.2 Write a simple run script
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#173-not-quite-ready" class="md-nav__link">
    <span class="md-ellipsis">
      
        1.7.3 Not quite ready!
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="17-running-nf-core-on-hpc">1.7 Running nf-core on HPC</h1>
<div class="admonition info">
<p class="admonition-title">Learning objectives</p>
<ul>
<li>Build a basic run script for a nf-core pipeline</li>
<li>Run an nf-core pipeline</li>
<li>Understand the need for HPC-specific configurations</li>
</ul>
</div>
<h2 id="171-an-overview-of-the-inputs">1.7.1 An overview of the inputs</h2>
<p>As mentioned in the <a href="./02_1_nfcore_intro.md">previous section</a>, we will be running a small section of the overall <code>sarek</code> workflow: the <code>mapping</code> stage. This stage requires a few distinct inputs:</p>
<ul>
<li>FASTQ files: These are plain text files that contain the sequences of the reads that came from the sequencing machine.</li>
<li>A <strong>samplesheet</strong>: A <code>.csv</code> file that describes each sample, its name, and where to find its associated FASTQ files.</li>
<li>A FASTA file: This holds the actual sequence of the chromosomes in the reference genome.</li>
<li>A BWA index: This is a group of files that is used by the alignment tool <code>bwa mem</code> to efficiently find where in the genome a particular sequencing read best maps to.</li>
</ul>
<p>We have two samplesheets prepared for today's workshop, both within the <code>../data/fqs</code> directory. The first is <code>samplesheet.fq.csv</code>. You can display it on the command line with:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>cat<span class="w"> </span>../data/fqs/samplesheet.fq.csv
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Viewing the samplesheet in VSCode</p>
<p>Unfortunately, because the <code>data/</code> folder is in the parent directory to where we are working, we can't see it in the VSCode explorer. However, if you wish to open the samplesheet in the VSCode editor, you can still do so via the terminal using the <code>code</code> command:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>code<span class="w"> </span>../data/fqs/samplesheet.fq.csv
</code></pre></div>
</div>
<p>The samplesheet looks like this:</p>
<div class="highlight"><span class="filename">samplesheet.fq.csv</span><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>patient,sample,lane,fastq_1,fastq_2
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>NA12877,test_sample1,all,../data/fqs/NA12877_chr20-22.R1.fq.gz,../data/fqs/NA12877_chr20-22.R2.fq.gz
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>NA12878,test_sample2,all,../data/fqs/NA12878_chr20-22.R1.fq.gz,../data/fqs/NA12878_chr20-22.R2.fq.gz
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>NA12889,test_sample3,all,../data/fqs/NA12889_chr20-22.R1.fq.gz,../data/fqs/NA12889_chr20-22.R2.fq.gz
</code></pre></div>
<p>The first line is a <strong>header line</strong>: it describes the "column names" of the table that the CSV represents. We have four columns in this file:</p>
<ul>
<li><code>patient</code>: An identifier for the individual the sample came from. Note that more than one sample may originate from an individual, such as in the case of paired tumor and normal samples.</li>
<li><code>sample</code>: A unique identifier for the sample. Each line must have a unique value for this column.</li>
<li><code>lane</code>: This identifies the lane of the flow cell that the sequencing read originated from.</li>
<li><code>fastq_1</code>: The path to the read 1 FASTQ file for that sample.</li>
<li><code>fastq_2</code>: The path to the read 2 FASTQ file for that sample.</li>
</ul>
<p>We can also see that we have three samples defined in our samplesheet. Samplesheets make it very easy to add new samples to our workflow run, simply by adding a new line and filling out the relevant columns.</p>
<p>The second samplesheet we have prepared is <code>samplesheet.single.csv</code>:</p>
<div class="highlight"><span class="filename">samplesheet.single.csv</span><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>patient,sample,lane,fastq_1,fastq_2
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>NA12877,test_sample1,all,../data/fqs/NA12877_chr20-22.R1.fq.gz,../data/fqs/NA12877_chr20-22.R2.fq.gz
</code></pre></div>
<p>As you can see, this contains just a single sample. For the inital part of this lesson, we will work with this samplesheet in order to keep things simple and quick.</p>
<div class="admonition note">
<p class="admonition-title">Samplesheet columns for sarek</p>
<p>The <code>sarek</code> pipeline can use a lot of different tools and start and stop at various stages. This means there are different columns required in the samplesheet depending on what you are running. Check with the <a href="https://nf-co.re/sarek/3.5.0/docs/usage/">usage documentation</a> to determine the correct columns for your run. In our case, we are using the <a href="https://nf-co.re/sarek/3.5.0/docs/usage/#start-with-mapping---step-mapping-default">minimal configuration</a> when starting with the <code>mapping</code> stage.</p>
</div>
<details class="example">
<summary>Additional content: A deeper look at the input files</summary>
<p>For those who are interested, here we look a bit closer at the structure of these input files.</p>
<h3 id="fastq-files">FASTQ files</h3>
<p>FASTQ files are plain text files that are structured in a particular way to capture the sequences of the reads in our sample as well as the quality of those sequences. Each read is represented by <strong>four lines</strong> in a FASTQ file:</p>
<div class="highlight"><span class="filename">Two reads from a FASTQ file</span><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>@HSQ1004:134:C0D8DACXX:1:1105:9389:51835
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>TGTGTTCCTGATTTGGCTCCTGGCTTGACTGTTGGTGTATAAGAATGTTAATGATTTTTGCACATTGATTTTGTATCCTGAGACTTTGCTGAAGTTATCAG
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>+
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>CCCFFFFFHHGHHJJJJJJJJJJJJJJIJJJGHIICG?FHGIJGIGIDGIIIIIIJJGIGGGIJJIJJIJJJJIIEHEFGGHFFDFFFEEDCEECDEECDD
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a>@HSQ1004:134:C0D8DACXX:1:2101:9406:116923
<a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a>AGGTTTTATTTTTAAATTTTTTTCTTTAGTTTTGTCTGACTGTGTTGATTCAAAGGACTGATCTTTGAGCTCTGAGATTCTTTCCTCAACTTGATCTATTC
<a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a>+
<a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a>@@?DBDDEBFFFFHIIIIDGHIIGEIIIG9DGGGIIGIIHGHFCBFC&gt;GHGHGHCGEGGGIIEGIG;@=EHCEE@?:BEDADE@@@@;@ACACDCC&gt;CD;@
</code></pre></div>
<p>Each set of four lines is defined as follows:</p>
<ol>
<li>The read name. This often includes information about the flow cell it came from.</li>
<li>The sequence of the read. This is the actual As, Cs, Gs and Ts of the read.</li>
<li>A <code>+</code> symbol. This line isn't normally used for anything and serves mostly as a visual separator between the sequence and its quality scores.</li>
<li>The quality scores for the read. The characters on this line represent quality scores for the corresponding position in the sequence. The quality scores are encoded as <a href="https://www.ascii-code.com/">ASCII characters</a>, starting at ASCII code 33 (<code>!</code>) for a value of <code>0</code> and increasing from there. The quality scores themselves are called <a href="https://en.wikipedia.org/wiki/Phred_quality_score"><strong>Phred quality scores</strong></a>, which are logarithmically related to the probability of an error at that base (higher scores being of better quality). This encoding scheme is referred to as <strong>Phred-33</strong>.</li>
</ol>
<p>We are working with <strong>paired-end</strong> sequencing data, where every sequencing read has two <strong>mates</strong> originating from either end of the original DNA fragment. These are typically separated into two ordered FASTQ files, one containing the first mates (or "read 1") and the other containing the second mates (or "read 2").</p>
<p>FASTQ files are also often compressed as <strong>gzip</strong> files. This helps to reduce storage space for large sequencing samples.</p>
<h3 id="fasta-files">FASTA files</h3>
<p>A FASTA file is another a plain text file type that holds the full sequence of the reference genome that the sequencing data will be aligned to. The FASTA format is much simpler than FASTQ. Each sequence starts with a line beginning with a <code>&gt;</code> character. This line provides the name for the sequence, e.g. <code>chr20</code>. The actual sequence starts on the following line and can be split into any number of lines. The sequence ends when either a new one begins (with a new line starting with <code>&gt;</code>) or the file ends.</p>
<div class="highlight"><span class="filename">An example FASTA file</span><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>&gt;seq1
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>ATCGGTCA
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>&gt;seq2
<a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a>AATTGGCC
<a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a>GGCCTTAA
<a id="__codelineno-5-6" name="__codelineno-5-6" href="#__codelineno-5-6"></a>&gt;seq3
<a id="__codelineno-5-7" name="__codelineno-5-7" href="#__codelineno-5-7"></a>AATGGATA
<a id="__codelineno-5-8" name="__codelineno-5-8" href="#__codelineno-5-8"></a>CTTGATAC
<a id="__codelineno-5-9" name="__codelineno-5-9" href="#__codelineno-5-9"></a>CCTG
</code></pre></div>
<p>The FASTA file we are working with today (<code>Hg38.subsetchr20-22.fasta</code>) contains three sequences: human chromosomes 20-22. These are the smallest autosomal chromosomes, and were picked for this workshop to ensure speedy alignment. You'll notice that the first 600 lines of the <code>chr20</code> sequence are just full of <code>N</code> characters; this is a "mask" character, and is used to essentially block out regions of the genome that are unreliably sequenced, as is the case with the chromosome ends. It isn't until line 602 that we see some actual A's, C's, G's and T's. In total, there are more than half a million lines in the file.</p>
<div class="highlight"><span class="filename">Hg38.subsetchr20-22.fasta</span><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>&gt;chr20
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>...
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a>ATTATCCATGAGGAGGGAGTGCAGACAAAGCAAAGAAGGAGGATGTTTGGAGAGGGGTAGTCTTTGAGTGGAGCCTTTAGGGATGAGAAGGGTGAATTGA
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a>GATATACCGGGAAAGTAGAAAAGATAAAACACGTATTAAAT
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a>&gt;chr21
<a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
<a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
</code></pre></div>
<p>The FASTA file is typically associated with two additional helper files: the FASTA index (<code>.fai</code>) and dictionary (<code>.dict</code>) files. Both contain information about the sequences in the main FASTA file and are used by tools for quickly finding sequences within it, especially when the genome is very large.</p>
<h3 id="bwa-index">BWA Index</h3>
<p>The job of <code>bwa mem</code> is to take sequencing reads in the FASTQ format and try to align them to a reference genome. For all but the most trivially-sized genomes, this is a very computationally-intensive task. Therefore, <code>bwa mem</code> makes use of an index to more efficiently look up where in the genome specific sequences map to. BWA's index format consists of a directory containing multiple files:</p>
<div class="highlight"><span class="filename">An example BWA index</span><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>Homo_sapiens_assembly38.20-22.fasta.amb  Homo_sapiens_assembly38.20-22.fasta.pac
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>Homo_sapiens_assembly38.20-22.fasta.ann  Homo_sapiens_assembly38.20-22.fasta.sa
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a>Homo_sapiens_assembly38.20-22.fasta.bwt
</code></pre></div>
<p>Note that different alignment tools will use their own specific index format; this format is specific to <code>bwa mem</code> and can't be used with other tools.</p>
</details>
<h2 id="172-write-a-simple-run-script">1.7.2 Write a simple run script</h2>
<p>To run the mapping stage of <code>sarek</code>, we need to execute the <code>nextflow run</code> command and include several parameters. We will use a run script to compose this command bit-by-bit to keep things organised.</p>
<div class="admonition example">
<p class="admonition-title">Exercise: Create a run script for nf-core/sarek</p>
<p>In the current working directory, create a new blank file and name it <code>run.sh</code>. You can do this via VSCode's interface by right-clicking on the current directory (<code>part1</code>) in the explorer side bar, clicking on "New File...", writing "run.sh" and hitting the Enter key. You can also do this via the terminal:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>touch<span class="w"> </span>run.sh
</code></pre></div>
<p>You should also make the file executable by running the following command:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>chmod<span class="w"> </span>+x<span class="w"> </span>run.sh
</code></pre></div>
<p>Next, we'll start the run command by adding the following lines:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">Gadi (PBSpro)</label><label for="__tabbed_1_2">Setonix (Slurm)</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><span class="filename">run.sh</span><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="ch">#!/bin/bash</span>
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a>
<a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a>module<span class="w"> </span>load<span class="w"> </span>nextflow/24.04.5
<a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a>module<span class="w"> </span>load<span class="w"> </span>singularity
<a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a>
<a id="__codelineno-10-6" name="__codelineno-10-6" href="#__codelineno-10-6"></a>nextflow<span class="w"> </span>run<span class="w"> </span>sarek/main.nf<span class="w"> </span><span class="se">\</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><span class="filename">run.sh</span><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="ch">#!/bin/bash</span>
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>
<a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a>module<span class="w"> </span>load<span class="w"> </span>nextflow/24.10.0
<a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a>module<span class="w"> </span>load<span class="w"> </span>singularity/4.1.0-slurm
<a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a>
<a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a>nextflow<span class="w"> </span>run<span class="w"> </span>sarek/main.nf<span class="w"> </span><span class="se">\</span>
</code></pre></div>
</div>
</div>
</div>
<p>So far, this is pretty straight-forward: we first tell the OS that we're running a bash script with the <code>#!/bin/bash</code> comment. Then, we load the nextflow and singularity modules; these are necessary for ensuring that the <code>nextflow</code> command is available to us, and that Nextflow is able to use Singularity for running containers. Finally, we run the <code>main.nf</code> Nextflow script within the <code>sarek</code> repository.</p>
<p>As it exists, this won't run: the <code>sarek</code> pipeline requires you to give it a samplesheet as input, and we also need to configure a few other parameters to ensure that we only run our small <code>mapping</code> stage of the pipeline. That's why we've ended the <code>nextflow run</code> line with a backslash (<code>\</code>), indicating that the command continues on the following line.</p>
<p>Add the following lines to define the inputs to the <code>mapping</code> stage:</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">run.sh</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-12-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-12-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-12-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-12-10">10</a></span>
<span class="normal"><a href="#__codelineno-12-11">11</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-12-7" name="__codelineno-12-7"></a><span class="w">    </span>--input<span class="w"> </span>../data/fqs/samplesheet.single.csv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-8" name="__codelineno-12-8"></a><span class="w">    </span>--fasta<span class="w"> </span>../data/ref/Hg38.subsetchr20-22.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-9" name="__codelineno-12-9"></a><span class="w">    </span>--fasta_fai<span class="w"> </span>../data/ref/Hg38.subsetchr20-22.fasta.fai<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-10" name="__codelineno-12-10"></a><span class="w">    </span>--dict<span class="w"> </span>../data/ref/Hg38.subsetchr20-22.dict<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-11" name="__codelineno-12-11"></a><span class="w">    </span>--bwa<span class="w"> </span>../data/ref<span class="w"> </span><span class="se">\</span>
</code></pre></div></td></tr></table></div>
<p>We've provided here the samplesheet CSV file to the <code>--input</code> parameter, as well as the FASTA file and its paired index (FAI) and dictionary files to the <code>--fasta</code>, <code>--fasta_fai</code>, and <code>--dict</code> parameters. We also specified the path to the BWA index directory with <code>--bwa</code>.</p>
<div class="admonition note">
<p class="admonition-title">Finding more information about the sarek parameters</p>
<p>The <a href="https://nf-co.re/sarek/3.5.0/parameters/">sarek parameters documentation</a> contains information about all of the parameters supported by <code>sarek</code> and the valid arguments to them. The <a href="https://nf-co.re/sarek/3.5.0/docs/usage/">usage page</a> in the docs also contains useful information about which parameters are required by the various stages of the pipeline.</p>
</div>
<p>Next, we'll specify that we want to run the pipeline from the <code>mapping</code> step, and skip several downstream tools, including <code>markduplicates</code>, <code>baserecalibrator</code>, <code>mosdepth</code>, and <code>samtools</code>. This will have the effect of just running the <code>mapping</code> step, followed by a final run of MultiQC; in total we will run only eight distinct processes, which will take only a few minutes for our dataset:</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">run.sh</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-13-12">12</a></span>
<span class="normal"><a href="#__codelineno-13-13">13</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-13-12" name="__codelineno-13-12"></a><span class="w">    </span>--step<span class="w"> </span>mapping<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-13-13" name="__codelineno-13-13"></a><span class="w">    </span>--skip_tools<span class="w"> </span>markduplicates,baserecalibrator,mosdepth,samtools<span class="w"> </span><span class="se">\</span>
</code></pre></div></td></tr></table></div>
<p>Wrapping up, we will tell the pipeline where to place our outputs, as well as configure a few additional parameters to make things run a bit smoother in our small example:</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">run.sh</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-14-14">14</a></span>
<span class="normal"><a href="#__codelineno-14-15">15</a></span>
<span class="normal"><a href="#__codelineno-14-16">16</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-14-14" name="__codelineno-14-14"></a><span class="w">    </span>--outdir<span class="w"> </span>results<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-14-15" name="__codelineno-14-15"></a><span class="w">    </span>--no_intervals<span class="w"> </span><span class="nb">true</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-14-16" name="__codelineno-14-16"></a><span class="w">    </span>--igenomes_ignore<span class="w"> </span><span class="nb">true</span>
</code></pre></div></td></tr></table></div>
<p>The <code>--no_intervals true</code> parameter tells <code>sarek</code> that we don't want to worry about splitting our data up into distinct genomic intervals. Intervals are a very useful feature for large datasets, as it lets us parallelise large genomic sequencing data into chunks that can be processed simultaneously - and takes advantage of the parallel nature of HPCs! However, in our very small example here, it would actually cause things to run slower by adding more jobs and waiting for them to start and finish.</p>
<p>The next line, <code>--igenomes_ignore true</code> stops the pipeline from downloading some additional files from public databases; again, this can be useful in a proper run, but for our purposes it is more of a nuisance.</p>
<p>At the end, the <code>run.sh</code> script should look like the following:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="2:2"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><div class="tabbed-labels"><label for="__tabbed_2_1">Gadi (PBSpro)</label><label for="__tabbed_2_2">Setonix (Slurm)</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">run.sh</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-15-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-15-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-15-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-15-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-15-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-15-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-15-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-15-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-15-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-15-10">10</a></span>
<span class="normal"><a href="#__codelineno-15-11">11</a></span>
<span class="normal"><a href="#__codelineno-15-12">12</a></span>
<span class="normal"><a href="#__codelineno-15-13">13</a></span>
<span class="normal"><a href="#__codelineno-15-14">14</a></span>
<span class="normal"><a href="#__codelineno-15-15">15</a></span>
<span class="normal"><a href="#__codelineno-15-16">16</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1"></a><span class="ch">#!/bin/bash</span>
<a id="__codelineno-15-2" name="__codelineno-15-2"></a>
<a id="__codelineno-15-3" name="__codelineno-15-3"></a>module<span class="w"> </span>load<span class="w"> </span>nextflow/24.04.5
<a id="__codelineno-15-4" name="__codelineno-15-4"></a>module<span class="w"> </span>load<span class="w"> </span>singularity
<a id="__codelineno-15-5" name="__codelineno-15-5"></a>
<a id="__codelineno-15-6" name="__codelineno-15-6"></a>nextflow<span class="w"> </span>run<span class="w"> </span>sarek/main.nf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-15-7" name="__codelineno-15-7"></a><span class="w">    </span>--input<span class="w"> </span>../data/fqs/samplesheet.single.csv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-15-8" name="__codelineno-15-8"></a><span class="w">    </span>--fasta<span class="w"> </span>../data/ref/Hg38.subsetchr20-22.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-15-9" name="__codelineno-15-9"></a><span class="w">    </span>--fasta_fai<span class="w"> </span>../data/ref/Hg38.subsetchr20-22.fasta.fai<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-15-10" name="__codelineno-15-10"></a><span class="w">    </span>--dict<span class="w"> </span>../data/ref/Hg38.subsetchr20-22.dict<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-15-11" name="__codelineno-15-11"></a><span class="w">    </span>--bwa<span class="w"> </span>../data/ref<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-15-12" name="__codelineno-15-12"></a><span class="w">    </span>--step<span class="w"> </span>mapping<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-15-13" name="__codelineno-15-13"></a><span class="w">    </span>--skip_tools<span class="w"> </span>markduplicates,baserecalibrator,mosdepth,samtools<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-15-14" name="__codelineno-15-14"></a><span class="w">    </span>--outdir<span class="w"> </span>results<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-15-15" name="__codelineno-15-15"></a><span class="w">    </span>--no_intervals<span class="w"> </span><span class="nb">true</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-15-16" name="__codelineno-15-16"></a><span class="w">    </span>--igenomes_ignore<span class="w"> </span><span class="nb">true</span>
</code></pre></div></td></tr></table></div>
</div>
<div class="tabbed-block">
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">run.sh</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-16-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-16-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-16-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-16-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-16-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-16-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-16-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-16-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-16-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-16-10">10</a></span>
<span class="normal"><a href="#__codelineno-16-11">11</a></span>
<span class="normal"><a href="#__codelineno-16-12">12</a></span>
<span class="normal"><a href="#__codelineno-16-13">13</a></span>
<span class="normal"><a href="#__codelineno-16-14">14</a></span>
<span class="normal"><a href="#__codelineno-16-15">15</a></span>
<span class="normal"><a href="#__codelineno-16-16">16</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1"></a><span class="ch">#!/bin/bash</span>
<a id="__codelineno-16-2" name="__codelineno-16-2"></a>
<a id="__codelineno-16-3" name="__codelineno-16-3"></a>module<span class="w"> </span>load<span class="w"> </span>nextflow/24.10.0
<a id="__codelineno-16-4" name="__codelineno-16-4"></a>module<span class="w"> </span>load<span class="w"> </span>singularity/4.1.0-slurm
<a id="__codelineno-16-5" name="__codelineno-16-5"></a>
<a id="__codelineno-16-6" name="__codelineno-16-6"></a>nextflow<span class="w"> </span>run<span class="w"> </span>sarek/main.nf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-7" name="__codelineno-16-7"></a><span class="w">    </span>--input<span class="w"> </span>../data/fqs/samplesheet.single.csv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-8" name="__codelineno-16-8"></a><span class="w">    </span>--fasta<span class="w"> </span>../data/ref/Hg38.subsetchr20-22.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-9" name="__codelineno-16-9"></a><span class="w">    </span>--fasta_fai<span class="w"> </span>../data/ref/Hg38.subsetchr20-22.fasta.fai<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-10" name="__codelineno-16-10"></a><span class="w">    </span>--dict<span class="w"> </span>../data/ref/Hg38.subsetchr20-22.dict<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-11" name="__codelineno-16-11"></a><span class="w">    </span>--bwa<span class="w"> </span>../data/ref<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-12" name="__codelineno-16-12"></a><span class="w">    </span>--step<span class="w"> </span>mapping<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-13" name="__codelineno-16-13"></a><span class="w">    </span>--skip_tools<span class="w"> </span>markduplicates,baserecalibrator,mosdepth,samtools<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-14" name="__codelineno-16-14"></a><span class="w">    </span>--outdir<span class="w"> </span>results<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-15" name="__codelineno-16-15"></a><span class="w">    </span>--no_intervals<span class="w"> </span><span class="nb">true</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-16" name="__codelineno-16-16"></a><span class="w">    </span>--igenomes_ignore<span class="w"> </span><span class="nb">true</span>
</code></pre></div></td></tr></table></div>
</div>
</div>
</div>
</div>
<div class="admonition question">
<p class="admonition-title">How are you going?</p>
<p>If you're following along so far, let us know by reacting on zoom with a <strong>"<span class="twemoji check"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 7 9 19l-5.5-5.5 1.41-1.41L9 16.17 19.59 5.59z"/></svg></span> Yes"</strong>.</p>
<p>If you're running into any issues, please react with a <strong>"<span class="twemoji close"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg></span> No"</strong> and we can help out before we move on to the next section.</p>
</div>
<h2 id="173-not-quite-ready">1.7.3 Not quite ready!</h2>
<p>We've created a full run script for our pipeline, but we won't run it just yet, because we haven't actually told Nextflow to use the HPC scheduler. If we were to run the workflow, it would attempt to run on the login node where we are currently working!</p>
<p>In addition, we haven't configured Nextflow to use Singularity yet, so Nextflow will assume all of the necessary software is installed and ready to use - which it isn't.</p>
<p>In the next section, we will build up a configuration file that tells Nextflow to both use Singularity and to talk to the HPC's scheduler to submit jobs to the compute nodes.</p>









  




                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../01_6_nfcore_intro/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 1.6 Intro to nf-core">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                1.6 Intro to nf-core
              </div>
            </div>
          </a>
        
        
          
          <a href="../01_8_nfcore_config/" class="md-footer__link md-footer__link--next" aria-label="Next: 1.8 Configuring nf-core">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                1.8 Configuring nf-core
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../..", "features": ["navigation.footer", "content.code.copy", "content.tabs.link"], "search": "../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
    
  </body>
</html>